File: 14.UnionFindSet并查集/leetcode_2492_minScore.py
from typing import List

class UFS:
    def __init__(self, n):
        self.pa = list(range(n))

    def find(self, x):
        if self.pa[x] != x:
            self.pa[x] = self.find(self.pa[x])
        return self.pa[x]
    
    def union(self, x, y):
        px, py = self.find(x), self.find(y)
        if px < py:
            self.pa[py] = px
        else:
            self.pa[px] = py

class Solution:
    def minScore(self, n: int, roads: List[List[int]]) -> int:
        ufs = UFS(n)
        for a,b,d in roads:
            a -= 1
            b -= 1
            ufs.union(a, b)
        
        target = ufs.find(0)
        dic = set()
        for i in range(n):
            if ufs.find(i) == target:
                dic.add(i)
        
        res = 10**18
        for a,b,d in roads:
            a -= 1
            b -= 1
            if a in dic and b in dic:
                res = min(res, d)
        return res

if __name__ == "__main__":
    s = Solution()
    n = 4
    roads = [[1,2,9],[2,3,6],[2,4,5],[1,4,7]]
    print(s.minScore(n, roads))

File: 14.UnionFindSet并查集/leetcode_2492_minScore.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include <climits>
using namespace std;

class UFS{
private:
    vector<int> pa;
public:
    UFS(int n){
        for (int i=0; i<n; i++){
            pa.push_back(i);
        }
    }

    int find(int x){
        if (pa[x] != x){
            pa[x] = find(pa[x]);
        }
        return pa[x];
    }

    void merge(int x, int y){
        int px = find(x), py = find(y);
        if (px<py){
            pa[py] = px;
        }
        else{
            pa[px] = py;
        }
    }
};

class Solution {
public:
    int minScore(int n, vector<vector<int>>& roads) {
        UFS ufs(n);
        for (auto& r : roads){
            int a=r[0]-1, b=r[1]-1, d=r[2];
            ufs.merge(a, b);
        }

        int target = ufs.find(0);
        unordered_set<int> st;
        for (int i=0; i<n; i++){
            if (ufs.find(i) == target){
                st.insert(i);
            }
        }

        int res = INT_MAX;
        for (auto& r : roads){
            int a=r[0]-1, b=r[1]-1, d=r[2];
            if (st.count(a) && st.count(b)){
                res = min(res, d);
            }
        }
        return res;
    }
};

int main(){
    Solution s;
    int n = 4;
    vector<vector<int>> roads = {{1,2,9},{2,3,6},{2,4,5},{1,4,7}};
    cout << s.minScore(n, roads) << endl;
}
